### ISSUE TITLE
Replacement for MDArray, potential bad memory leak in Shards from this

### ISSUE THREAD

**Issue by lxmota** [gahansen/Albany#268]
@ibaned @ikalash @gahanse @agsalin While chasing a frustrating bug in Schwarz about resetting state arrays, I discovered that MDArrays are shards::Arrays, as defined in Albany_StateInfoStruct.hpp.

shards::Arrays overload the operator= with a shallow copy, hence the failure in resetting state variables for Schwarz.

There are two problems with this:

- There appears to be no way to do a deep copy of an MDArray/shard:Array, which is needed for Schwarz.
- The shallow copy uses raw pointers. It does not free the old pointer before overwriting it with the new one, hence the memory it was pointing to is lost.

My question is: is it possible to replace the MDArray in Albany_StateInfoStruct.hpp with something else? A Kokkos view perhaps?

I could jump through hoops and make this work with MDArrays somehow, but it seems to me that there should be a cleaner solution and also avoid the potential for memory leaks.

I'm open to suggestions.

**Comment by bartgol** [comment]
I thought shards::Array was going to be removed in some future release of Trilinos anyways, no?

Anyhow, Kokkos::view is a good candidate for replacement, I guess.

**Comment by mperego** [comment]
I would also replace it with Kokkos::view. Actually, since we do not necessarily know the rank at compile time we have to use Kokkos::DynRankView.

**Comment by bartgol** [comment]
Why don't we know the rank? State arrays have a fixed rank, it's only the dimension that is runtime. No?

**Comment by lxmota** [comment]
Yes, it appears that MDArrays are of rank 1, and that all the interval variables are stuffed linearly in them regardless of whether there are scalars, vectors, second order tensors, etc.

**Comment by bartgol** [comment]
Does it need to be that way though? Can't we enforce the rank at compile time, and have scalar, vector and tensor states, like we already do in the stk meshes fields?

**Comment by lxmota** [comment]
Yes, I guess one could do that. First I need to figure out how to solve the issue of the shallow copy and memory leak to have something working. But I'll keep that in mind, because that would be ideal instead of packing and unpacking things.

**Comment by ibaned** [comment]
I would start with replacing them with 1D `Kokkos::View`s, and move on from there. Small steps are usually the best way to go.

**Comment by lxmota** [comment]
@ibaned I quite agree.
